### ISSUE TITLE
Terms have http://purl.obolibrary.org/obo/TEMP# as URI

### ISSUE THREAD

**Issue by neobernad** [EBISPOT/OLS#242]
Hello,

I'm running a local installation of OLS. In my `ols-config.yaml` file I have set the *URI* and *BASE_URI* as:
```
uri: http://example.com/ontology/test
base_uri:
   - http://example.com/ontology/test#
```

However my terms are stored as http://purl.obolibrary.org/obo/TEMP#identifier, when I expected it to be set as http://example.com/ontology/test#identifier. What am I doing wrong?

Thanks,
José

**Comment by henrietteharmse** [comment]
Thank you for reporting Jose. 
I will look into this shortly.

**Comment by henrietteharmse** [comment]
Hi Jose,

Please can you provide the ontology you get this problem with.

Thanks,
Henriette

**Comment by neobernad** [comment]
Hello Henriette,

I cannot provide the actual ontology as it is since it is private, but I have created an example ontology where it occurs the same problem, it is an obo file:
```
format-version: 1.2
date: 01:01:2000 00:00
saved-by: test
synonymtypedef: SYNONYM "SYNONYM"
synonymtypedef: ABBR "Abrreviation"
remark: based on test_old.obo

[Term]
id: 1
name: Term 1
synonym: "syn1 term 1" EXACT SYNONYM [TEST] {date=2000-01-01, lang=en}
synonym: "syn2 term 1" EXACT SYNONYM [TEST] {date=2000-01-01, lang=en}
synonym: "syn3 term 1" EXACT SYNONYM [TEST] {date=2000-01-01, lang=en, prefname=1}
is_a: 1 ! Term

[Term]
id: 2
name: Term 2
synonym: "syn1 term 2" EXACT SYNONYM [TEST] {date=2018-05-09, lang=en}
synonym: "syn2 term 2" EXACT SYNONYM [TEST] {date=2018-05-09, lang=en, prefname=1}
synonym: "syn3 term 2" EXACT SYNONYM [TEST] {date=2018-05-09, lang=en}
is_a: 2 ! Term

[Term]
id: 3
name: Term 3
synonym: "syn1 term 3" EXACT SYNONYM [TEST] {date=2018-05-09, lang=en}
synonym: "syn2 term 3" EXACT SYNONYM [TEST] {date=2018-05-09, lang=en, prefname=1}
synonym: "syn3 term 3" EXACT SYNONYM [TEST] {date=2018-05-09, lang=en}
is_a: 3 ! Special term
```

The `ols-config.yaml` is the following:
```
# test.obo
  - id: id_test
    preferredPrefix: test
    title: Test title
    uri: http://example.com/ontology/test
    description: "Test description."
    base_uri:
      - http://example.com/ontology/test#
    oboSlims: false
    ontology_purl: http://localhost/ontology/test.obo
```

The ontology IRI is set to `http://example.com/ontology/test` correctly, but the terms have the *temporary* URI. For example 'Term 2' has `http://purl.obolibrary.org/obo/TEMP#2`. Maybe some metadata missing in our obo file?

Thank you for your time,
José Antonio

**Comment by henrietteharmse** [comment]
Hi Jose,

Thanks for supplying the example ontology.

Can you please change your ontology to be in OWL format rather than OBO format?

The reason for this is that OLS uses the OWL API internally which results in all terms of an ontology in OBO format to default to the OBO idspace. Currently there is no way around this besides using the OWL format. See [this OWL API issue](https://github.com/owlcs/owlapi/issues/753) for more details.

As for the `base_uri` config you supplied in the `ols-config.yaml` - that is used internally in OLS to determine terms that belong to your ontology versus terms that belong to other ontologies. Hence, it has *no* effect on the URIs of your ontology.

I hope that helps.

Henriette

**Comment by neobernad** [comment]
Hello Henriette,

Thanks for your time, we will pay attention to the OWL API and to the issue you have attached for a future fix.

Best regards,
José Antonio

**Comment by dosumis** [comment]
If your starting point is OBO, the this should work:

```yaml

format-version: 1.2
date: 01:01:2000 00:00
saved-by: test
synonymtypedef: SYNONYM "SYNONYM"
synonymtypedef: ABBR "Abrreviation"
remark: based on test_old.obo
ontology: http://example.com/ontology/test.owl

[Term]
id: http://example.com/ontology/test#1
name: Term 1
synonym: "syn1 term 1" EXACT SYNONYM [TEST] {date=2000-01-01, lang=en}
synonym: "syn2 term 1" EXACT SYNONYM [TEST] {date=2000-01-01, lang=en}
synonym: "syn3 term 1" EXACT SYNONYM [TEST] {date=2000-01-01, lang=en, prefname=1}

[Term]
id: http://example.com/ontology/test#2
name: Term 2
synonym: "syn1 term 2" EXACT SYNONYM [TEST] {date=2018-05-09, lang=en}
synonym: "syn2 term 2" EXACT SYNONYM [TEST] {date=2018-05-09, lang=en, prefname=1}
synonym: "syn3 term 2" EXACT SYNONYM [TEST] {date=2018-05-09, lang=en}
is_a: http://example.com/ontology/test#1 ! Term 1

[Term]
id: http://example.com/ontology/test#3
name: Term 3
synonym: "syn1 term 3" EXACT SYNONYM [TEST] {date=2018-05-09, lang=en}
synonym: "syn2 term 3" EXACT SYNONYM [TEST] {date=2018-05-09, lang=en, prefname=1}
synonym: "syn3 term 3" EXACT SYNONYM [TEST] {date=2018-05-09, lang=en}
is_a: http://example.com/ontology/test#2 ! Term 2

```

Here's what it looks like in owl (rdf-xml) after conversion using [Robot](http://robot.obolibrary.org/):

```xml
<?xml version="1.0"?>
<rdf:RDF xmlns="http://example.com/ontology/test.owl#"
     xml:base="http://example.com/ontology/test.owl"
     xmlns:owl="http://www.w3.org/2002/07/owl#"
     xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
     xmlns:xml="http://www.w3.org/XML/1998/namespace"
     xmlns:xsd="http://www.w3.org/2001/XMLSchema#"
     xmlns:rdfs="http://www.w3.org/2000/01/rdf-schema#"
     xmlns:test="http://purl.obolibrary.org/obo/http://example.com/ontology/test.owl#"
     xmlns:oboInOwl="http://www.geneontology.org/formats/oboInOwl#">
    <owl:Ontology rdf:about="http://example.com/ontology/test.owl">
        <oboInOwl:date rdf:datatype="http://www.w3.org/2001/XMLSchema#string">01:01:2000 00:00</oboInOwl:date>
        <oboInOwl:hasOBOFormatVersion rdf:datatype="http://www.w3.org/2001/XMLSchema#string">1.2</oboInOwl:hasOBOFormatVersion>
        <oboInOwl:saved-by rdf:datatype="http://www.w3.org/2001/XMLSchema#string">test</oboInOwl:saved-by>
        <rdfs:comment rdf:datatype="http://www.w3.org/2001/XMLSchema#string">based on test_old.obo</rdfs:comment>
    </owl:Ontology>
    


    <!-- 
    ///////////////////////////////////////////////////////////////////////////////////////
    //
    // Annotation properties
    //
    ///////////////////////////////////////////////////////////////////////////////////////
     -->

    


    <!-- http://purl.obolibrary.org/obo/http://example.com/ontology/test.owl#ABBR -->

    <owl:AnnotationProperty rdf:about="http://purl.obolibrary.org/obo/http://example.com/ontology/test.owl#ABBR">
        <rdfs:label rdf:datatype="http://www.w3.org/2001/XMLSchema#string">Abrreviation</rdfs:label>
        <rdfs:subPropertyOf rdf:resource="http://www.geneontology.org/formats/oboInOwl#SynonymTypeProperty"/>
    </owl:AnnotationProperty>
    


    <!-- http://purl.obolibrary.org/obo/http://example.com/ontology/test.owl#SYNONYM -->

    <owl:AnnotationProperty rdf:about="http://purl.obolibrary.org/obo/http://example.com/ontology/test.owl#SYNONYM">
        <rdfs:label rdf:datatype="http://www.w3.org/2001/XMLSchema#string">SYNONYM</rdfs:label>
        <rdfs:subPropertyOf rdf:resource="http://www.geneontology.org/formats/oboInOwl#SynonymTypeProperty"/>
    </owl:AnnotationProperty>
    


    <!-- http://www.geneontology.org/formats/oboInOwl#SynonymTypeProperty -->

    <owl:AnnotationProperty rdf:about="http://www.geneontology.org/formats/oboInOwl#SynonymTypeProperty">
        <rdfs:label rdf:datatype="http://www.w3.org/2001/XMLSchema#string">synonym_type_property</rdfs:label>
    </owl:AnnotationProperty>
    


    <!-- http://www.geneontology.org/formats/oboInOwl#date -->

    <owl:AnnotationProperty rdf:about="http://www.geneontology.org/formats/oboInOwl#date"/>
    


    <!-- http://www.geneontology.org/formats/oboInOwl#hasDbXref -->

    <owl:AnnotationProperty rdf:about="http://www.geneontology.org/formats/oboInOwl#hasDbXref">
        <rdfs:label rdf:datatype="http://www.w3.org/2001/XMLSchema#string">database_cross_reference</rdfs:label>
    </owl:AnnotationProperty>
    


    <!-- http://www.geneontology.org/formats/oboInOwl#hasExactSynonym -->

    <owl:AnnotationProperty rdf:about="http://www.geneontology.org/formats/oboInOwl#hasExactSynonym">
        <rdfs:label rdf:datatype="http://www.w3.org/2001/XMLSchema#string">has_exact_synonym</rdfs:label>
    </owl:AnnotationProperty>
    


    <!-- http://www.geneontology.org/formats/oboInOwl#hasOBOFormatVersion -->

    <owl:AnnotationProperty rdf:about="http://www.geneontology.org/formats/oboInOwl#hasOBOFormatVersion">
        <rdfs:label rdf:datatype="http://www.w3.org/2001/XMLSchema#string">has_obo_format_version</rdfs:label>
    </owl:AnnotationProperty>
    


    <!-- http://www.geneontology.org/formats/oboInOwl#hasSynonymType -->

    <owl:AnnotationProperty rdf:about="http://www.geneontology.org/formats/oboInOwl#hasSynonymType">
        <rdfs:label rdf:datatype="http://www.w3.org/2001/XMLSchema#string">has_synonym_type</rdfs:label>
    </owl:AnnotationProperty>
    


    <!-- http://www.geneontology.org/formats/oboInOwl#id -->

    <owl:AnnotationProperty rdf:about="http://www.geneontology.org/formats/oboInOwl#id"/>
    


    <!-- http://www.geneontology.org/formats/oboInOwl#lang -->

    <owl:AnnotationProperty rdf:about="http://www.geneontology.org/formats/oboInOwl#lang"/>
    


    <!-- http://www.geneontology.org/formats/oboInOwl#prefname -->

    <owl:AnnotationProperty rdf:about="http://www.geneontology.org/formats/oboInOwl#prefname"/>
    


    <!-- http://www.geneontology.org/formats/oboInOwl#saved-by -->

    <owl:AnnotationProperty rdf:about="http://www.geneontology.org/formats/oboInOwl#saved-by"/>
    


    <!-- http://www.w3.org/2000/01/rdf-schema#label -->

    <owl:AnnotationProperty rdf:about="http://www.w3.org/2000/01/rdf-schema#label"/>
    


    <!-- 
    ///////////////////////////////////////////////////////////////////////////////////////
    //
    // Classes
    //
    ///////////////////////////////////////////////////////////////////////////////////////
     -->

    


    <!-- http://example.com/ontology/test#1 -->

    <owl:Class rdf:about="http://example.com/ontology/test#1">
        <oboInOwl:hasExactSynonym rdf:datatype="http://www.w3.org/2001/XMLSchema#string">syn1 term 1</oboInOwl:hasExactSynonym>
        <oboInOwl:hasExactSynonym rdf:datatype="http://www.w3.org/2001/XMLSchema#string">syn2 term 1</oboInOwl:hasExactSynonym>
        <oboInOwl:hasExactSynonym rdf:datatype="http://www.w3.org/2001/XMLSchema#string">syn3 term 1</oboInOwl:hasExactSynonym>
        <oboInOwl:id rdf:datatype="http://www.w3.org/2001/XMLSchema#string">http://example.com/ontology/test#1</oboInOwl:id>
        <rdfs:label rdf:datatype="http://www.w3.org/2001/XMLSchema#string">Term 1</rdfs:label>
    </owl:Class>
    <owl:Axiom>
        <owl:annotatedSource rdf:resource="http://example.com/ontology/test#1"/>
        <owl:annotatedProperty rdf:resource="http://www.geneontology.org/formats/oboInOwl#hasExactSynonym"/>
        <owl:annotatedTarget rdf:datatype="http://www.w3.org/2001/XMLSchema#string">syn1 term 1</owl:annotatedTarget>
        <oboInOwl:date rdf:datatype="http://www.w3.org/2001/XMLSchema#string">2000-01-01</oboInOwl:date>
        <oboInOwl:hasDbXref rdf:datatype="http://www.w3.org/2001/XMLSchema#string">TEST</oboInOwl:hasDbXref>
        <oboInOwl:hasSynonymType rdf:resource="http://purl.obolibrary.org/obo/http://example.com/ontology/test.owl#SYNONYM"/>
        <oboInOwl:lang rdf:datatype="http://www.w3.org/2001/XMLSchema#string">en</oboInOwl:lang>
    </owl:Axiom>
    <owl:Axiom>
        <owl:annotatedSource rdf:resource="http://example.com/ontology/test#1"/>
        <owl:annotatedProperty rdf:resource="http://www.geneontology.org/formats/oboInOwl#hasExactSynonym"/>
        <owl:annotatedTarget rdf:datatype="http://www.w3.org/2001/XMLSchema#string">syn2 term 1</owl:annotatedTarget>
        <oboInOwl:date rdf:datatype="http://www.w3.org/2001/XMLSchema#string">2000-01-01</oboInOwl:date>
        <oboInOwl:hasDbXref rdf:datatype="http://www.w3.org/2001/XMLSchema#string">TEST</oboInOwl:hasDbXref>
        <oboInOwl:hasSynonymType rdf:resource="http://purl.obolibrary.org/obo/http://example.com/ontology/test.owl#SYNONYM"/>
        <oboInOwl:lang rdf:datatype="http://www.w3.org/2001/XMLSchema#string">en</oboInOwl:lang>
    </owl:Axiom>
    <owl:Axiom>
        <owl:annotatedSource rdf:resource="http://example.com/ontology/test#1"/>
        <owl:annotatedProperty rdf:resource="http://www.geneontology.org/formats/oboInOwl#hasExactSynonym"/>
        <owl:annotatedTarget rdf:datatype="http://www.w3.org/2001/XMLSchema#string">syn3 term 1</owl:annotatedTarget>
        <oboInOwl:date rdf:datatype="http://www.w3.org/2001/XMLSchema#string">2000-01-01</oboInOwl:date>
        <oboInOwl:hasDbXref rdf:datatype="http://www.w3.org/2001/XMLSchema#string">TEST</oboInOwl:hasDbXref>
        <oboInOwl:hasSynonymType rdf:resource="http://purl.obolibrary.org/obo/http://example.com/ontology/test.owl#SYNONYM"/>
        <oboInOwl:lang rdf:datatype="http://www.w3.org/2001/XMLSchema#string">en</oboInOwl:lang>
        <oboInOwl:prefname rdf:datatype="http://www.w3.org/2001/XMLSchema#string">1</oboInOwl:prefname>
    </owl:Axiom>
    


    <!-- http://example.com/ontology/test#2 -->

    <owl:Class rdf:about="http://example.com/ontology/test#2">
        <rdfs:subClassOf rdf:resource="http://example.com/ontology/test#1"/>
        <oboInOwl:hasExactSynonym rdf:datatype="http://www.w3.org/2001/XMLSchema#string">syn1 term 2</oboInOwl:hasExactSynonym>
        <oboInOwl:hasExactSynonym rdf:datatype="http://www.w3.org/2001/XMLSchema#string">syn2 term 2</oboInOwl:hasExactSynonym>
        <oboInOwl:hasExactSynonym rdf:datatype="http://www.w3.org/2001/XMLSchema#string">syn3 term 2</oboInOwl:hasExactSynonym>
        <oboInOwl:id rdf:datatype="http://www.w3.org/2001/XMLSchema#string">http://example.com/ontology/test#2</oboInOwl:id>
        <rdfs:label rdf:datatype="http://www.w3.org/2001/XMLSchema#string">Term 2</rdfs:label>
    </owl:Class>
    <owl:Axiom>
        <owl:annotatedSource rdf:resource="http://example.com/ontology/test#2"/>
        <owl:annotatedProperty rdf:resource="http://www.geneontology.org/formats/oboInOwl#hasExactSynonym"/>
        <owl:annotatedTarget rdf:datatype="http://www.w3.org/2001/XMLSchema#string">syn1 term 2</owl:annotatedTarget>
        <oboInOwl:date rdf:datatype="http://www.w3.org/2001/XMLSchema#string">2018-05-09</oboInOwl:date>
        <oboInOwl:hasDbXref rdf:datatype="http://www.w3.org/2001/XMLSchema#string">TEST</oboInOwl:hasDbXref>
        <oboInOwl:hasSynonymType rdf:resource="http://purl.obolibrary.org/obo/http://example.com/ontology/test.owl#SYNONYM"/>
        <oboInOwl:lang rdf:datatype="http://www.w3.org/2001/XMLSchema#string">en</oboInOwl:lang>
    </owl:Axiom>
    <owl:Axiom>
        <owl:annotatedSource rdf:resource="http://example.com/ontology/test#2"/>
        <owl:annotatedProperty rdf:resource="http://www.geneontology.org/formats/oboInOwl#hasExactSynonym"/>
        <owl:annotatedTarget rdf:datatype="http://www.w3.org/2001/XMLSchema#string">syn2 term 2</owl:annotatedTarget>
        <oboInOwl:date rdf:datatype="http://www.w3.org/2001/XMLSchema#string">2018-05-09</oboInOwl:date>
        <oboInOwl:hasDbXref rdf:datatype="http://www.w3.org/2001/XMLSchema#string">TEST</oboInOwl:hasDbXref>
        <oboInOwl:hasSynonymType rdf:resource="http://purl.obolibrary.org/obo/http://example.com/ontology/test.owl#SYNONYM"/>
        <oboInOwl:lang rdf:datatype="http://www.w3.org/2001/XMLSchema#string">en</oboInOwl:lang>
        <oboInOwl:prefname rdf:datatype="http://www.w3.org/2001/XMLSchema#string">1</oboInOwl:prefname>
    </owl:Axiom>
    <owl:Axiom>
        <owl:annotatedSource rdf:resource="http://example.com/ontology/test#2"/>
        <owl:annotatedProperty rdf:resource="http://www.geneontology.org/formats/oboInOwl#hasExactSynonym"/>
        <owl:annotatedTarget rdf:datatype="http://www.w3.org/2001/XMLSchema#string">syn3 term 2</owl:annotatedTarget>
        <oboInOwl:date rdf:datatype="http://www.w3.org/2001/XMLSchema#string">2018-05-09</oboInOwl:date>
        <oboInOwl:hasDbXref rdf:datatype="http://www.w3.org/2001/XMLSchema#string">TEST</oboInOwl:hasDbXref>
        <oboInOwl:hasSynonymType rdf:resource="http://purl.obolibrary.org/obo/http://example.com/ontology/test.owl#SYNONYM"/>
        <oboInOwl:lang rdf:datatype="http://www.w3.org/2001/XMLSchema#string">en</oboInOwl:lang>
    </owl:Axiom>
    


    <!-- http://example.com/ontology/test#3 -->

    <owl:Class rdf:about="http://example.com/ontology/test#3">
        <rdfs:subClassOf rdf:resource="http://example.com/ontology/test#2"/>
        <oboInOwl:hasExactSynonym rdf:datatype="http://www.w3.org/2001/XMLSchema#string">syn1 term 3</oboInOwl:hasExactSynonym>
        <oboInOwl:hasExactSynonym rdf:datatype="http://www.w3.org/2001/XMLSchema#string">syn2 term 3</oboInOwl:hasExactSynonym>
        <oboInOwl:hasExactSynonym rdf:datatype="http://www.w3.org/2001/XMLSchema#string">syn3 term 3</oboInOwl:hasExactSynonym>
        <oboInOwl:id rdf:datatype="http://www.w3.org/2001/XMLSchema#string">http://example.com/ontology/test#3</oboInOwl:id>
        <rdfs:label rdf:datatype="http://www.w3.org/2001/XMLSchema#string">Term 3</rdfs:label>
    </owl:Class>
    <owl:Axiom>
        <owl:annotatedSource rdf:resource="http://example.com/ontology/test#3"/>
        <owl:annotatedProperty rdf:resource="http://www.geneontology.org/formats/oboInOwl#hasExactSynonym"/>
        <owl:annotatedTarget rdf:datatype="http://www.w3.org/2001/XMLSchema#string">syn1 term 3</owl:annotatedTarget>
        <oboInOwl:date rdf:datatype="http://www.w3.org/2001/XMLSchema#string">2018-05-09</oboInOwl:date>
        <oboInOwl:hasDbXref rdf:datatype="http://www.w3.org/2001/XMLSchema#string">TEST</oboInOwl:hasDbXref>
        <oboInOwl:hasSynonymType rdf:resource="http://purl.obolibrary.org/obo/http://example.com/ontology/test.owl#SYNONYM"/>
        <oboInOwl:lang rdf:datatype="http://www.w3.org/2001/XMLSchema#string">en</oboInOwl:lang>
    </owl:Axiom>
    <owl:Axiom>
        <owl:annotatedSource rdf:resource="http://example.com/ontology/test#3"/>
        <owl:annotatedProperty rdf:resource="http://www.geneontology.org/formats/oboInOwl#hasExactSynonym"/>
        <owl:annotatedTarget rdf:datatype="http://www.w3.org/2001/XMLSchema#string">syn2 term 3</owl:annotatedTarget>
        <oboInOwl:date rdf:datatype="http://www.w3.org/2001/XMLSchema#string">2018-05-09</oboInOwl:date>
        <oboInOwl:hasDbXref rdf:datatype="http://www.w3.org/2001/XMLSchema#string">TEST</oboInOwl:hasDbXref>
        <oboInOwl:hasSynonymType rdf:resource="http://purl.obolibrary.org/obo/http://example.com/ontology/test.owl#SYNONYM"/>
        <oboInOwl:lang rdf:datatype="http://www.w3.org/2001/XMLSchema#string">en</oboInOwl:lang>
        <oboInOwl:prefname rdf:datatype="http://www.w3.org/2001/XMLSchema#string">1</oboInOwl:prefname>
    </owl:Axiom>
    <owl:Axiom>
        <owl:annotatedSource rdf:resource="http://example.com/ontology/test#3"/>
        <owl:annotatedProperty rdf:resource="http://www.geneontology.org/formats/oboInOwl#hasExactSynonym"/>
        <owl:annotatedTarget rdf:datatype="http://www.w3.org/2001/XMLSchema#string">syn3 term 3</owl:annotatedTarget>
        <oboInOwl:date rdf:datatype="http://www.w3.org/2001/XMLSchema#string">2018-05-09</oboInOwl:date>
        <oboInOwl:hasDbXref rdf:datatype="http://www.w3.org/2001/XMLSchema#string">TEST</oboInOwl:hasDbXref>
        <oboInOwl:hasSynonymType rdf:resource="http://purl.obolibrary.org/obo/http://example.com/ontology/test.owl#SYNONYM"/>
        <oboInOwl:lang rdf:datatype="http://www.w3.org/2001/XMLSchema#string">en</oboInOwl:lang>
    </owl:Axiom>
</rdf:RDF>



<!-- Generated by the OWL API (version 4.5.6) https://github.com/owlcs/owlapi -->

```

**Comment by neobernad** [comment]
Hello David, @dosumis

I tried to use `robot` to transform from OBO to OWL. Even though the IRIs of the classes are properly generated, some others are not. For example the *Annotation property* ABBR, it is created with the following IRI: `http://purl.obolibrary.org/obo/http://example.com/ontology/test.owl#ABBR` instead of `http://example.com/ontology/test.owl#ABBR`.

The only walkaround I found is opening the OBO ontology in Protègè, changing the Ontology IRI there, and save as RDF/XML, which is a quite slow procedure when you have a lot of ontologies to change...

Thanks for spending time on this issue,
José Antonio
